src/storage/memtable/ob_memtable_block_row_scanner.cpp

```diff
@@ -0,0 +1,223 @@
+/**
+ * Copyright (c) 2021 OceanBase
+ * OceanBase CE is licensed under Mulan PubL v2.
+ * You can use this software according to the terms and conditions of the Mulan PubL v2.
+ * You may obtain a copy of Mulan PubL v2 at:
+ *          http://license.coscl.org.cn/MulanPubL-2.0
+ * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
+ * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
+ * See the Mulan PubL v2 for more details.
+ */
+
+#include "ob_memtable_block_row_scanner.h"
+#include "storage/access/ob_block_row_store.h"
+#include "storage/blocksstable/ob_row_reader.h"
+
+namespace oceanbase
+{
+namespace memtable
+{
+
+int ObMemtableBlockRowScanner::init(
+    common::ObIAllocator *allocator,
+    const int64_t capacity,
+    char *trans_info_ptr,
+    const storage::ObITableReadInfo *read_info)
+{
+  int ret = OB_SUCCESS;
+  allocator_ = allocator;
+  read_info_ = read_info;
+  for (int i = 0; i < ObMemtableBlockRowScanner::MAX_ROW_BATCH_SIZE; ++i) {
+    if (OB_FAIL(rows_[i].init(*allocator_, capacity, trans_info_ptr))) {
+      STORAGE_LOG(WARN, "Fail to init memtable scan row batch", K(ret), K(i));
+      break;
+    }
+  }
+  return ret;
+}
+
+blocksstable::ObDatumRow *ObMemtableBlockRowScanner::get_next_row(bool fast_filter_skipped)
+{
+  rows_[cur_idx_].fast_filter_skipped_ = fast_filter_skipped;
+  return rows_ + (cur_idx_++);
+}
+
+bool ObMemtableBlockRowScanner::is_valid()
+{
+  bool ret = false;
+  if (OB_ISNULL(result_bitmap_)) {
+    ret = (cur_idx_ < batch_size_);
+  } else if (cur_idx_ < batch_size_) {
+    int iter = cur_idx_;
+    for (; iter < batch_size_ && !result_bitmap_->test(iter); ++iter) {}
+    if (iter < batch_size_) {
+      cur_idx_ = iter;
+      ret = true;
+    } else {
+      ret = false;
+    }
+  } else {
+    ret = false;
+  }
+  return ret;
+}
+
+int ObMemtableBlockRowScanner::filter_pushdown_filter(
+    sql::ObPushdownFilterExecutor *parent,
+    sql::ObPushdownFilterExecutor *filter,
+    storage::PushdownFilterInfo &pd_filter_info,
+    common::ObBitmap &bitmap)
+{
+  int ret = OB_SUCCESS;
+  blocksstable::ObStorageDatum *col_buf = pd_filter_info.datum_buf_;
+  // const int64_t col_capacity = pd_filter_info.col_capacity_;
+  // TODO(jianxian): validate_filter_info.
+  int64_t col_count = filter->get_col_count();
+  const common::ObIArray<int32_t> &col_offsets = filter->get_col_offsets();
+  const sql::ColumnParamFixedArray &col_params = filter->get_col_params();
+  const common::ObIArray<blocksstable::ObStorageDatum> &default_datums = filter->get_default_datums();
+  const blocksstable::ObColDescIArray &cols_desc = read_info_->get_columns_desc();
+
+  for (int64_t row_idx = pd_filter_info.start_; OB_SUCC(ret) && row_idx < pd_filter_info.end_; ++row_idx) {
```

Comment:
If pd_filter_info.start_ is always 0, perhaps we could use pd_filter_info.row_count_ to iterate  all rows? 

Use row_count appears to be clearer and more straightforward to comprehend for me.